rfcs/0005-version-matrix.md

```diff
@@ -0,0 +1,477 @@
+These scenarios ramp up in frequency very fast in our low traffic projects like Druid or Piet.
+In cases where a [simple one line fix PR](https://github.com/linebender/piet/pull/561) could be accepted we're instead greeted with the task of having to deal with the accumulated decay of the last 6 months.
+
+In similar vein it would be great if contributors could run various tests locally in a manner that approximates the CI, to cut down on surprising errors later.
+
+### Historic build reproduction
+
+Sometimes you want to check out an older revision of a project and compile it.
+For example if there is a newly found bug, but it was introduced a while ago and you want to debug it in the version with the least changes.
+It would be very useful indeed to be able to compile this historic revision without having to start dealing with the fact that some unknown sub-dependency's newer version isn't compatible and causes unexpected further issues.
+
+### We want to support more than just the latest stable Rust toolchain
+
+Not everyone can or will upgrade their Rust toolchain the moment a new version is published.
+This includes potential new contributors, to whom we want to let write their patch without having to start updating their toolchain.
+Beyond that, it will be even more common among application developers who use our libraries.
+It's a question of balance of course, but we want to support toolchains as old as possible.
+
+### We want a widely applicable strategy
+
+In the absence of a well thought out general strategy we will have (potentially ephemeral) ad hoc approaches for every distinct project.
+This means an increased maintenance burden, as problems solved with one approach won't necessarily carry over to another.
+Also there will be more problems to solve, because a mix of strategies will bring us the union of their respective problems.
+
+
+## User-facing explanation
+
+We test and support only the latest versions of all our package dependencies.
+
+Our minimum supported Rust version (MSRV) is largely decided by our dependencies and we just document the fact.
+
+
+## Implementation strategy
+
+### Keep the existing general Linebender CI as a base and extend it
+
+Our current general CI script is the result of years of experience.
+It already has good techniques to avoid some sources of dependency contamination, e.g. feature activations caused by dev dependencies.
+It also already pins the stable Rust toolchain version, allowing us to choose when to update it.
+Those benefits should move forward and as such this RFC doesn't see the need to remove anything and all changes are additions.
+
+### `Cargo.lock` is committed to source control
+
+Having a one true `Cargo.lock` solves a lot of the contributor and CI issues.
+Contributors can easily use the same dependency versions as the CI does.
+The CI won't suddenly start producing errors just because a new dependency version was published that causes issues.
```

Comment:
I feel like it would be better if CI did start producing errors in such circumstances. That way we'll know about them and can fix the issue. Most users will be using the latest versions of everything (as noted above), so it's important that this works.

Reply:
We would lose the benefits I outline in the RFC, most notably contributor friction and historic build reproduction. Ultimately I think the CI is a tool meant to help the developers of the project, and so taking trade offs in the developer's favor makes sense.

We still care about these potential breakages. That is why the RFC advocates for regular updates, to not let a single `Cargo.lock` setup linger for too long.

Also, crucially, the RFC advocates for updating the dependencies just before publishing. So these issues wouldn't be present for users in the immediate time window after release.

Users who already depend on one of our libraries have their own `Cargo.lock` with a working configuration. If they run `cargo update` and then their build fails, they can easily revert the `Cargo.lock` changes.

Thus the only scenario where this would be a real problem is when a user adds one of our libraries as a new dependency after enough time has passed from our release date that one of our sub dependencies is now problematic. But also critically this has to be before one of our regular `Cargo.lock` updates so that we haven't caught the issue and dealt with it already.

This is about finding a balance, and after careful consideration I belive that commiting `Cargo.lock` with a regular update cadence is the right trade-off to make. It minimizes the time window where issues could appear for users while greatly improving the contributor experience.

Similar conclusions are being reached in the wider Rust ecosystem and indeed just last year the [Cargo team changed their guidance around `Cargo.lock`](https://blog.rust-lang.org/2023/08/29/committing-lockfiles.html) to reflect that many in the ecosystem are now committing it for good reason.

On the downsides front, I think that the potential merge conflicts will be a much bigger source of concern for us than users finding broken builds.